### PR TITLE
build(gradle): Kotlin DSL build and Dependabot

### PR DESCRIPTION
**Pull request by stachu540**
### Prerequisites for Code Changes
* [x] This pull request follows the code style of the project
* [x] I have tested this feature

### Changes Proposed

1. Revmap build to `kotlin-dsl`
2. creating `.github/dependabot.yml` for dependencies

### Additional Information 

Revmaped build to Kotlin DSL will help simplify build library. Creating `buildSrc` and provide function and his extensions gives oportunity to doing some cleanup. Of course created `gradle.properties` provides only `group` and `version`. If you wish extracting version from environmental variables i can move it to `buildSrc/src/main/kotlin/_globals.kt` which are having more environmental variables definition into property when they are accessible for all projects. All changes are tested without no issue (for now). **WARNING**: *Do not add `apply false` into `plugins` definition. It cause exceptions of missing properties or functions, which are be not loaded if you didn't apply plugin inside the file. More about cross configuration project will find [here](https://docs.gradle.org/current/userguide/kotlin_dsl.html#sec:kotlin_cross_project_configuration)*

Dependabot is supported for [Kotlin DSL](https://github.com/dependabot/dependabot-core/pull/2680). Preview version doesn't have it this feature.

After resolving #244 i will start resolve this PR. For now marking him as Draft.

### CODE REVIEW

**Comment by stachu540** [comment]
I'll hold kotlin dsl changes. The bug is more complex that we can imagine it. Extended functions defined for dependencies will be break it, if you try fetching jar resources classes (I think). To avoid that we will use a quoted configuration like that: `"implementation"(project(":common"))`. This feature provided from shadow plugin breaks a configurations defined in kotlin-dsl's extended functions.

Simple: Using kotlin-dsl with `shadow` pluigin and define [this lines](https://github.com/twitch4j/twitch4j/blob/8e01ecfd70e88e3e4a470ead833be0d9763ae2a0/deployment.gradle#L28-L46) ,and will disallow loading extended functions, resolving errors of unknown extended function. This should be fixed only quoting configuration name as mentioned above.

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		withType<Jar> {
+			archiveBaseName.set(artifactId)
+		}
+
+		val delombok by getting(io.freefair.gradle.plugins.lombok.tasks.Delombok::class)
+		val jar by getting(Jar::class)
+		val relocateShadowJar by creating(com.github.jengelman.gradle.plugins.shadow.tasks.ConfigureShadowRelocation::class) {
+			target = shadowJar.get()
+			prefix = "com.github.twitch4j.shaded.${"$version".replace(".", "_")}"
+		}
```
Comment (stachu540) [inline build.gradle.kts:132]: I have use this task. [Reference](https://github.com/johnrengelman/shadow/issues/629)

File: buildSrc/src/main/kotlin/globals.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+enum class VersionType(private val pattern: Regex) {
+	MAJOR(Regex("(\\d+).\\d+.\\d+-.*")),
+	MINOR(Regex("\\d+.(\\d+).\\d+-.*")),
+	PATCH(Regex("\\d+.\\d+.(\\d+)-.*"));
+
+	internal fun replace(version: String) = version.replace(pattern, "${"$1".toInt() + 1}")
+}
+
+fun nextIterable(version: String, type: VersionType, snapshot: Boolean = true) =
+	type.replace(version) + (if (snapshot) "-SNAPSHOT" else "")
```
Comment (stachu540) [inline buildSrc/src/main/kotlin/globals.kt:34]: This should be removed after approved

File: buildSrc/src/main/kotlin/project.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+fun MavenPomDeveloperSpec.PhilippHeuer() {
+	developer {
+		id.set("PhilippHeuer")
+		name.set("Philipp Heuer")
+		email.set("git@philippheuer.me")
+		roles.addAll("creator", "developer", "owner")
+	}
+}
```
Comment (stachu540) [inline buildSrc/src/main/kotlin/project.kt:44]: Whole POM configuration moved here. Each subprojects have this sample line:
```kotlin
publishing.publications.withType<MavenPublication> {
	pom {
		name.set("Twitch4J Chat Module") // project title
		description.set("Chat dependency") // project description
	}
}
```

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7.1-all.zip
```
Comment (stachu540) [inline gradle/wrapper/gradle-wrapper.properties:3]: This should be bumped to 6.8

**Comment by stachu540** [comment]
This is ready to review and deploy. Only some typings (spaces, new-lines, etc.).

File: gradle.properties [hunk 5]
```diff
@@ -0,0 +1,2 @@
+group=com.github.twitch4j
+version=1.2.0-SNAPSHOT
```
Comment (stachu540) [inline gradle.properties:2]: Almost forgot this. `groupId` and `version` moved to this file. [This plugin](https://github.com/KengoTODA/gradle-semantic-release-plugin/) works better if version will be here.

File: build.gradle.kts [hunk 6]
```diff
@@ -1,37 +1,27 @@
-@file:Suppress("UnstableApiUsage")
-
 // Plugins
 plugins {
-	signing
+	`signing`
```
Comment (stachu540) [inline build.gradle.kts:3]: no need backtick quoting that
```suggestion
	signing
```

File: build.gradle.kts [hunk 7]
```diff
@@ -44,7 +34,7 @@ subprojects {
 	apply(plugin = "com.github.johnrengelman.shadow")
 
 	lombok {
-		version.set("1.18.16")
+		version.set("1.18.18")
```
Comment (stachu540) [inline build.gradle.kts:37]: Lombok newer version
```suggestion
		version.set("1.18.20")
```

**Comment by stachu540** [comment]
I put some suggestion comments here. Everything is ready to review.

**Comment by stachu540** [comment]
I have add [manifest plugin](https://github.com/coditory/gradle-manifest-plugin) to this branch.

This is how looks like in my machine

```mf
Manifest-Version: 1.0
Implementation-Title: twitch4j-common
SCM-Commit-Message: chore: adding manifest plugin to auto-generate
Implementation-Version: unspecified
SCM-Commit-Hash: 856591ac3ea28800976cfd8ec6921d07fc06d4cc
Built-By: stachu
Built-Date: 2021-04-20T03:01:37Z
SCM-Commit-Author: Damian <359222+stachu540@users.noreply.github.com>
Built-Host: ESCANOR
Built-JDK: 1.8.0_281 Oracle Corporation
Implementation-Group: twitch4j
SCM-Branch: refs/heads/build/gradle-kts-dependabot
SCM-Repository: https://github.com/stachu540/twitch4j.git
Built-OS: Windows 10 10.0 amd64
SCM-Commit-Date: 2021-04-20T03:01:03Z
```

Will be added in all jars because this plugin adds only add for `jar` task

File: build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+			api(group = "com.github.vladimir-bukhtoyarov", name = "bucket4j-core", version = "4.7.0")
+
+			// HTTP
+			api(group = "com.squareup.okhttp3", name = "okhttp", version = "4.9.1")
+
+			// Event Dispatcher
+			api(group = "com.github.philippheuer.events4j", name = "events4j-core", version = "0.9.5")
+			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-simple", version = "0.9.5")
+			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-spring", version = "0.9.5")
+			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-reactor", version = "0.9.5")
```
Comment (iProdigy) [inline build.gradle.kts:76]:
```suggestion
			api(group = "com.github.philippheuer.events4j", name = "events4j-core", version = "0.9.6")
			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-simple", version = "0.9.6")
			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-spring", version = "0.9.6")
			api(group = "com.github.philippheuer.events4j", name = "events4j-handler-reactor", version = "0.9.6")
```

**Review by PhilippHeuer — APPROVED** [review]

**Review by iProdigy — APPROVED** [review]
Thanks for your hard work on this!

**Review by stachu540 — COMMENTED** [review]
`artifactId` will be setted in [here](https://github.com/twitch4j/twitch4j/pull/243/files#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R38-R40)

**Comment by PhilippHeuer** [comment]
> `artifactId` will be setted in [here](https://github.com/twitch4j/twitch4j/pull/243/files#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R38-R40)

Yes, but the generated artifactId's from the module names don't match the existing artifact names we published in the past, so i used fixed values everywhere to make sure it's pushed correctly.
For example the artifactId for helix should be twitch4j-helix, while the module name rest-helix would end up as twitch4j-rest-helix.

**Pull request merged**